### PR TITLE
End-to-end infrastructure bootstrap 

### PR DESCRIPTION
**Pull request by stefanprodan**
This PR adds e2e testing for Flux Helm chart using [flux-get-started](https://github.com/weaveworks/flux-get-started) repo:
- install Kubernetes Kind
- install Helm and Tiller
- install a git server and clone the flux-get-started repo 
- build and load the Flux container image into Kind
- install Flux and helm-op with the local helm chart
- sync the flux-get-started repo with Flux
- wait for the demo namespace to be created
- check the rollout status for workloads and Helm releases

Things to consider as a followup to this PR:
- load the Flux Helm-op local image into Kind
- use the local image when installing the helm release
- enable registry scanning and test the automated update

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -48,3 +57,7 @@ workflows:
           filters:
             tags:
               only: /(helm-)?[0-9]+(\.[0-9]+)*(-[a-z]+)?/
+      - e2e-testing:
+          filters:
+            tags:
+              only: /(helm-)?[0-9]+(\.[0-9]+)*(-[a-z]+)?/
```
Comment (squaremo) [inline .circleci/config.yml]: newline please! unless it upsets CircleCI?

File: test/e2e/git-dep.yaml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  namespace: flux
+spec:
+  ports:
+    - name: ssh
+      port: 22
+      protocol: TCP
+      targetPort: ssh
+  selector:
+    name: gitsrv
+  type: ClusterIP
```
Comment (squaremo) [inline test/e2e/git-dep.yaml]: newline please

File: .circleci/config.yml (hunk 1 above)
Comment (squaremo) [reply .circleci/config.yml]: Is the filter there to be conservative about when the end-to-end testing is run, for now?

File: test/e2e/known_hosts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+gitsrv ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDt65M0saFA6Ewsjc8mQ405bI406BsFSJrbGz9GFrbWBn0Rs3Loi4S6ADWgTWlcspHvbfmf7vXW9oiTjxeke8oHPCfFYrHFdPK26Big2J1kE2DzRONizZWcdw8dpqRhvxlt2+/EJuuay8CGXz3VLCV8NgJc5AYmAwNPknUhWKxaAzjwvRd/0cyXr4vvcTDcmwR63oiWcRPkHCZ5LClFuZE065Ulm2o6CgItl+e6M5ouDSJWZDpYWWmmJJJv1DQoRruNbaf5f4bgWVkKjrQ/0cA4iWWlkwJMlAWQgp9saD0G2F83hrf2XapM/clWTFybkzPqPqaw2BEwXP6vWp6A1iUR
+gitsrv ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBM6WyUxQRPtZhYLxjG6YRQOHD/SX/+uI4XBo44UTSu21uqmf/lG8cLWTdMVzDlUDY9/Dx4tFz96LT97kUC1pLJs=
+gitsrv ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIAMv1z2YnDug5M4Klp3FMbBvw9NdFrx7Om5uEKFQs07t
```
Comment (squaremo) [inline test/e2e/known_hosts]: newline please (unless it upsets SSH)

File: .circleci/config.yml (hunk 1 above)
Comment (squaremo) [reply .circleci/config.yml]: Actually this doesn't seem to filter what CircleCi does -- the e2e-testing jobs is run for this PR. So I have misunderstood what this means. What does it mean?

File: test/e2e/ssh-key.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
...aGJueHJaT0wxY3VhT1lWQUJnWWh5b2JnMWVabE10Mk1pVgpVaWxaamFkRnZ5emZjR1BTVEpjSjVmMmkvQ0dzS243RXFxUXdoZnd1S3ZRZTQyazgwbk5BcUczdTkvdkl4bklCCnFGZW5kTTE3SlN2WkU3NXFCVE9uTXVVZ1NuNFJoTXpzOEg3UTFmZFQ4UGMvTVRmRVVKcTQzcGtDZ1lFQXpGOUMKd1g0Z0UvUnZlbWZRc0ZVZ29TQ0lQdkR1bFhNNzNsK09XRnByOE00MWxpeGcvd0lvQ2NFZlpGcFdLUkpqSmwvUwpOVFh3YVhnOGg4RGl3a3d3dzJmcmNvWTl2TGNIcGxvWVRkN1ZjUVk4UGRKdjNJeGFReld6SHpMR3N0M29hZ08rCmJDbStsMEY5TnY0VUdWRHUrT0RSQjJyRWo2b1ZGRmh0SUQxbmRtVUNnWUJHS3V3alQrMkFzZlFSM2F1Q1p4eloKcVFDWmhBajM3QWEwV1RXOENhUE1UYUhrSUJ3VUtHN3FxUHRKaWlicngyNnAzbzRaMTU2QVNVemdrd1h3Y1lhaQptQUtKSHkrdHVtb1ZvcGdZTzE2Mzh5LzkrSGt1N3hCellZQmpwV3JGTEUxaHF6SGVFOFFnejREbm56ZUtrb2QxCmZLOWp5UUZMR1hDQXhSNGg1bGpES1FLQmdRQytqUjlmNjZvYkVQQ1Q3NUhicHpPS0tCd0FtNEhJWkszd2M2WHoKNlRMMVRqOFdhd0J4SStDUzM3YldTWWhHT1RlckF2S3EzRVR4QWNObVM4amhva3BoRjFhbTdGVkp6Rm5jbCtwTApTTFkzOExsZ1p3SVhYK0dWQXMrbENpSExpaTMyRXRHTVpndW5XYzlXNCtWM2lVZVhVMzV4N1BHaWhkR3JxNXJyCjBYVFRKUUtCZ0FReUF0RlloVHRONktCSER2NFdiTDQxcnBtcUlVcUlpV0R6a3FPT1ZXcHgzYkpTdWVNeDEyUjQKWHVVaGkwL2ZqbGFvMmYwWTBqbTBDUlQ5ZmlhQW56WHNMRXNzN2JYQ0ZZcGt3V3ZrNnNqV1BCWGdPUnBZbklHNQpRRWNFeklzRDFKQm1EY0RxdWxpZ0dnUzNIdGhiWTl5WW4vU3l4d0owcU5ob3BDS1d2OWNOCi0tLS0tRU5EIFJTQSBQUklWQVRFIEtFWS0tLS0tCg==
```
Comment (squaremo) [inline test/e2e/ssh-key.yaml]: :grimacing: can this be generated during setup?

File: .circleci/config.yml (hunk 1 above)
Comment (stefanprodan) [reply .circleci/config.yml]: Yeah that filter does nothing, CircleCI config so fun 😞 

Looks like a branch filter with exclude all is required:

```yaml
  filters:
    tags:
      only: /(helm-)?[0-9]+(\.[0-9]+)*(-[a-z]+)?/
    branches:
      ignore: /.*/
```

**Comment by stefanprodan** [comment]
I feel like this PR is useless without building and loading the Flux local image into Kind. With CircleCI you can't share images across jobs unless you pay for it.

**Comment by stefanprodan** [comment]
The only solution I could think if is to use a single `machine` job. Install golang and run the whole pipeline on that machine.

**Comment by stefanprodan** [comment]
@squaremo I'm happy now with the e2e as it uses the Flux image built locally to run the tests.

File: .circleci/config.yml (hunk 1 above)
Comment (hiddeco) [reply .circleci/config.yml]: Removing this will disable builds for tags. https://circleci.com/docs/2.0/workflows/#executing-workflows-for-a-git-tag

Comment (stefanprodan) [reply .circleci/config.yml]: Ahhh ok I'll put it back

File: test/e2e/ssh-key.yaml (hunk 4 above)
Comment (hiddeco) [reply test/e2e/ssh-key.yaml]: +1 for this, only thing holding me back from approval.

**Comment by 2opremio** [comment]
I am not too happy about rebuilding flux for the end to end tests. Also I think that the end to end tests could be run in parallel with the unit tests.

I will see what I can do.

**Comment by hiddeco** [comment]
> I am not too happy about rebuilding flux for the end to end tests. Also I think that the end to end tests could be run in parallel with the unit tests.

I do agree with you on both points but I think getting this merged now - as is - would already be of great value, any improvement can happily follow (shortly) after imho.

**Comment by 2opremio** [comment]
Alright, let's merge it then.

**Pull request merged**